core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,695 @@
+                        'tokens': list(str). Token list of program.
+                    } for each program in training_data.
+                }
+        token_to_id: dict. A dictionary which maps each token to a unique ID.
+        K: int. Model parameter 'K' of winnowing.
+
+    Returns:
+        dict. A dictionary containing training data with additional key
+        'k_gram_hashes' for each data point, storing k-gram hashes corresponding
+        to that data point. Structure of dict is as follows,
+            {
+                ID: {
+                    'source': str. Source code of program.
+                    'class': int. Associated feedback class with program.
+                    'tokens': list(str). Token list of program.
+                    'k_gram_hashes': list(int). K-gram hash values of program.
+                } for each program in training_data.
+            }
+    """
+    for program_id in training_data:
+        training_data[program_id]['k_gram_hashes'] = (
+            winnowing.k_gram_hash_generator(
+                training_data[program_id]['tokens'], token_to_id, K))
+    return training_data
+
+
+def generate_program_fingerprints(training_data, T, K):
+    """Generate document fingerprints for all programs in entire dataset.
+
+    Args:
+        training_data: dict. A dictionary containing training data. Structure of
+            training_data is as follows,
+                {
+                    ID: {
+                        'source': str. Source code of program.
+                        'class': int. Associated feedback class with program.
+                        'tokens': list(str). Token list of program.
+                        'k_gram_hashes': list(int). K-gram hash values of
+                            program.
+                    } for each program in training_data.
+                }
+        T: int. Model parameter 'T' of winnowing.
+        K: int. Model parameter 'K' of winnowing.
+
+    Returns:
+        dict. A dictionary containing training data with additional key
+        'fingerprint' for each program which stores generated fingerprint
+        of program. Structure of dict is as follows,
+            {
+                ID: {
+                    'source': str. Source code of program.
+                    'class': int. Associated feedback class with program.
+                    'tokens': list(str). Token list of program.
+                    'k_gram_hashes': list(int). K-gram hash values of program.
+                    'fingerprint': list(int). Extracted fingerprint of program.
+                } for each program in training_data.
+            }
+    """
+    window_size = T - K + 1
+    for program_id in training_data:
+        training_data[program_id]['fingerprint'] = (
+            winnowing.get_fingerprint_from_hashes(
+                training_data[program_id]['k_gram_hashes'], window_size))
+    return training_data
+
+
+def calc_jaccard_index(set_a, set_b):
```

Comment:
Ok, got you! But in the general programming sense (at least in the Python world), I wouldn't strictly call it a set because it still contains duplicate elements. I totally understand that what you want is correct and acceptable (in fact if you use set operations, you won't be able to capture the correct coefficient of similarity). In that case, could you please indicate in the comment that the set is not strictly a Python set but in fact a list which could contain duplicate.

Also, I see that you are using brute force approach of looping through both the sets. Perhaps, you can use sorting instead.